ably/ably-chat-js
# react hooks use room

### Context

* [CHA-382](https://ably.atlassian.net/browse/CHA-382)
* [CHADR-47](https://ably.atlassian.net/wiki/spaces/CHA/pages/3249602575/draft+CHADR-047+React+RoomProvider+attach+and+release+behaviour)

### Description

Adds `useRoom` hook, `<ChatRoomProvider>` provider and a `ChatRoomContext`.

In this version the `ChatRoomProvider` has `attach` and `release` props, both defaulting to `true`, to control whether this provider should automatically attach/detach (mount/unmount) or release the room (on unmount). As part of CHADR-47 this might change. The current thinking is to add a reference count and always auto-attach the room when the provider first provider mounts and always auto-release when the last provider unmounts.

### Checklist

Status: adding last few tests, nothing major should change otherwise.


### Testing Instructions (Optional)

* See how `useRoom`, `ChatRoomContext` are used in the demo app.
* `npm run test`
* read the code

[CHA-382]: https://ably.atlassian.net/browse/CHA-382?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
